### PR TITLE
RFC: Makes salt minion's master_port and publish_port options depend on each master's ip

### PR DESCRIPTION
**Pull request by kstreee**
- Feature Name: Makes salt minion's master_port and publish_port options depend on each master's ip.
- Start Date: 2018-12-12
- RFC PR:
- Salt Issue:

# Summary
[summary]: #summary

For example, the current salt minion configuration can't cover the following setting of multi-master.

master1's config
```
 interface: 1.1.1.1
 ret_port: 30001
 publish_port: 30002
```
master2's config
```
 interface: 1.1.1.2
 ret_port: 30008
 publish_port: 30009
```

minion's config
```
master:
  1.1.1.1:30001
  1.1.1.2:30001
publish_port: ?
```

# Motivation
[motivation]: #motivation

We are trying to deploy salt master using kubernetes, and kubernetes-like systems allocate network resources highly dynamically to use the limited resources efficiently. To leverage the system's network environment (dynamically allocate ports to expose container's ports to the outer network), `publish_port` and `ret_port` of salt masters of a multi-master environment can be assigned as different ports like the above setting.

# Design
[design]: #detailed-design

Current salt minion's master config
```
master:
  1.1.1.1:RET_PORT (MASTER_PORT)
  1.1.1.2:RET_PORT (MASTER_PORT)
publish_port: PUBLISH_PORT
```

Desired salt minion's master config
```
master_uri_format: connection_dict
master:
  - host: 1.1.1.1
    master_port: 30001
    publish_port: 30002
  - host: 1.1.1.2
    master_port: 30008
    publish_port: 30009
```

## Alternatives
[alternatives]: #alternatives

Uses nested to parse ip, master port, publish port, (inspired by @isbm's comment)

```
master_uri_format: nested
master:
- master: [2001:0db8:85a3:0000:0000:8a2e:0370:7334]:3001
  publish_port: 3002
- master: 123.123.123.123:3001
  publish_port: 3002
- master: 123.123.123.123  # master port will be set as 4506 (default port)
  publish_port: 3002
  master_uri_format: ip_only
- master: 123.123.123.123  # master port will be set as 4506, publish port will be set as 4505 (default port)
```

## Unresolved questions
[unresolved]: #unresolved-questions

Format of the new `master_uri_format` option. should it be nested style? or should it be simple structured dictionary style? is there any other option to represent this kinds of information better?

# Drawbacks
[drawbacks]: #drawbacks

- Backward compatibility, how to support both formats (old format, and new format) of master connection information of salt minion config?
- Implementation cost, we should find all being affected points in the minion code, and patch it to accept the new form of salt master connection config.
- Documentation

Adding a new `master_uri_format` is a simple job for the current version of minion. The only point that we should modify is the function `prep_ip_port`.
Make the function can handle this new `master_uri_format`, then it will work well without any further modifications.

------------

### What does this PR do?

Adds a new option for 'master_uri_format', and 'master' config format to make minion can handle more flexible master connection information.
Following two parts are unfixed points yet, so please give some opinions.
1. I don't know when this feature going to be available (versionadded field for this feature in the doc.. 2018.?.?) https://github.com/saltstack/salt/pull/50521/files#diff-bd9f376b73bc6dc98548702377da3791R70
2. new feature's option name. 'connection_dict' is a temporal naming. If any better naming for this feature, let me know to change it 😂.

### What issues does this PR fix or reference?

#50412

### Previous Behavior

Can't handle the case that each master (in a multi-master environment) has different ports for `master_port` and `publish_port`. More detailed description is in #50412.

### New Behavior

Can handle the case through the new type of master config. i.e.
```
master:
- host: 1.1.1.1
  master_port: 4506
  publish_port: 4505
- host: 1.1.1.2
  master_port: 34506
  publish_port: 34505
```

### Tests written?

Yes

### Commits signed with GPG?

Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/minion.py [hunk 1]
```diff
@@ -241,26 +242,43 @@ def resolve_dns(opts, fallback=True):
...
+    # Master connectino information has been configured as a dictionary type of data.
+    # i.e. {'host': 'x.x.x.x', 'master_port': y, 'publish_port': z}
+    if opts['master_uri_format'] == 'connection_dict':
+        if not isinstance(opts['master'], dict) \
+           or 'host' not in opts['master'] or not opts['master']['host'] \
+           or 'master_port' not in opts['master'] or not opts['master']['master_port'] \
+           or 'publish_port' not in opts['master'] or not opts['master']['publish_port']:
+            raise SaltConfigurationError('The option "master_uri_format" is set as connection_dict,'
+                                         ' but "master" is configured inappropriately')
+	ret['master'] = opts['master']['host']
```
Comment (cachedout) [inline salt/minion.py]: Indentation here is wrong.

**Comment by austinpapp** [comment]
yes please 🌮 👍

File: salt/minion.py (hunk 1 above)
Comment (kstreee) [reply salt/minion.py]: I fixed it..! thanks for the review. 👍

**Review by isbm — CHANGES_REQUESTED** [review]
Could you please first make RFC and PR it so the design is first discussed and finished, and _then_ implement this? I'd recomment either `WP` this PR or even close it for now. RFCs aren't issues, because during the discussion you are going to make changes to the RFC and then it becomes a documentation, once merged. Hope, this won't be a problem.

Let's discuss your RFC first, because e.g. I have few questions of the design overall.

**Comment by kstreee** [comment]
I thought about the comment @isbm left at #50412, and 'nested' could work for what @isbm want and options and code integrated smoothly with previous options.

i.e.
```
master_uri_format: nested
master:
  - master: [2001:0db8:85a3:0000:0000:8a2e:0370:7334]:3001
     publish_port: 3002
  - master: 123.123.123.123:3001
     publish_port: 3002
  - master: 123.123.123.123  # master port will be set as 4506 (default port)
     publish_port: 3002
     master_uri_format: ip_only
  - master: 123.123.123.123  # master port will be set as 4506, publish port will be set as 4505 (default port)
   ...
```

Code and options will be more intuitive than the current version of pr, i.e.

```
def prep_ip_port(opts):
    # ...
    if opts['master_uri_format'] == 'nested']
        opts.update({
            'master': opts['master']['master'],
            'master_port': opts['master']['master_port'],
           ..
        })
        return prep_ip_port(oprs)
```

**Review by isbm — CHANGES_REQUESTED** [review]
There is still a problem with this PR and I would be sorry asking you to either close it or rework this one as POC PR. :wink: To properly do an RFC PR, you should:

- [Go here](https://github.com/saltstack/salt/tree/develop/rfcs) in your Git clone
- Grab a copy of `0000-template.md` and name it to your topic.
- Open it in `$EDITOR` and write all that stuff there
- You may reference your this POC PR

But I would not recommend you to code something unless you're playing, because as we go reviewing and discussing RFC, this may dramatically change or even render as wrong approach. Or not, but you never know. :wink: 

Once your RFC PR is merged and your doc lands in [here](https://github.com/saltstack/salt/tree/develop/rfcs), then you may start working on an actual implementation. And then reuse your this RFC doc as an actual documentation for the release.

Hope this helps!

**Comment by kstreee** [comment]
@isbm @cachedout Now I guess, I finally understand what RFC process exactly is 😂😂😂 Thanks for the guiding me. I will close this, and make a correct RFC PR for this feature. Thanks all!

**Comment by isbm** [comment]
@kstreee you could still incorporate the comments here in your future RFC. Looking forward to re-read it.

**Comment by kstreee** [comment]
I reopened this issue, and removed the previous pr (code), and adds a rfc file. Sorry for the misunderstanding about RFC. 

@isbm I checked your comment, but minions can't have 'default' options for connection information for masters. I think you might be talking about master's, but this rfc is only for minion's config.

**Comment by isbm** [comment]
> @isbm I checked your comment, but minions can't have 'default' options for connection information for masters. I think you might be talking about master's, but this rfc is only for minion's config.

Sure they can. What if you pre-build a Docker image or VM image with different ports there pre-installed, so that is because your master is running on other ports?

Also I am not sure you understand the purpose of "default" section. It is to override system default. E.g., Salt default is 4505 and 4506. And so if you write in default "5505 and 5506", then everywhere else in your multu-master/cluster setup do not need to mention that anymore. It will be "default" to 5505/5506. Likewise you do not need the whole "default" section at all, if you want that default ports would be set to 4505/4506. 

To recap:

1. System default
2. "default" overrides system default
3. Particular section overrides "default".

**Comment by kstreee** [comment]
@isbm
Hi, sorry for late the response. I was busy for awhile 😂..

To my understanding,
```
master:
  - host: '[2001:0db8:85a3:0000:0000:8a2e:0370:7334]'  # master_port will be 123123 (salt's default)
    publish_port: 3002
  - host: 123.123.123.123 # master_port will be 123123 (salt's default)
    publish_port: 3002
  - host: foo.bar.blah.com # publish_port will be 1111 (non salt's default)
    master_port: 3001
  - default:  # non salt's default
      publish_port: 1111
  publish_port: 10101 # will be ignored (salt's default) <- then, when will be used?
  master_port: 123123 # will be used (salt's default)
```

Is it right?
I think that the semantics of non Salt's default options and Salt's default options in minion config is completely same. If non Salt default has been set, then every master contact points will follow the non Salt's default options and ignore Salt's default options. In that case, may I get an example of some logic that use the Salt's default options (master_port, publish_port) even when non Salt's default options has been set?

**Comment by dwoz** [comment]
@kstreee We have standardized the RFC process and it has become "Salt Enhancement Proposals". Can you please create a PR for this one In the new repository and link back to the discussion here?
https://github.com/saltstack/salt-enhancement-proposals/pulls

**Comment by dwoz** [comment]
@kstreee I am this for now. If you still would like to move this forward, please open a new PR [here](https://github.com/saltstack/salt-enhancement-proposals/pulls) and include a link back to this issue.

**Comment by kstreee** [comment]
@dwoz I missed the alarm. I will create a pr for this issue as you said. Thanks!

**Comment by nunesvn** [comment]
@kstreee Hey, did you created that PR, or do you know if now there is any new way to achieve this result? (Different publish_port per master on multimaster)
